### ISSUE TITLE
Crate relies on implicit promotion of `const fn` calls

### ISSUE THREAD

**Issue by RalfJung** [rodrimati1992/abi_stable_crates#46]
There are plans to [change the rules for implicit promotion](https://github.com/rust-lang/rfcs/pull/3027) such that calls to `const fn` are not implicitly promoted to `'static` lifetime any more. A [crater experiment](https://github.com/rust-lang/rust/pull/80243) showed that only very few crates would be affected by this change, and this is one of them.

You can see an example of the relevant build failure [here](https://crater-reports.s3.amazonaws.com/pr-80243/try%2381275a1445c7502a974118e0c69b795ab60be991/reg/plugin_tls-0.2.3/log.txt). Usually there is a way to work around this by explicitly putting the intermediate result into a `const`, of which a reference is then taken. However, since this is macro-generated code, I was unable to find the right spot(s) in the code where this would have to be done.

**Comment by rodrimati1992** [comment]
:frowning: This would be two orders of magnitude more work than every other time that the compiler decided to break my code combined.

**Comment by rodrimati1992** [comment]
So is there an easy way to get the nightly compiler with these changes on x86-64 linux to try making those changes?

**Comment by rodrimati1992** [comment]
I wouldn't have as much of a problem with this change if it was easier to declare `const`ants that can access the generic parameters in the surrounding scope, the only way to do that today is with associated constants.

**Comment by RalfJung** [comment]
> they reference generic parameters

Ouch, yeah that makes it quite annoying. :/
Note that nothing here is decided yet -- the RFC got accepted, but the RFC explicitly said that we'll do crater experiments to determine the fallout of rejecting such code, and explore alternatives if the fallout is considered too big. This kind of feedback is very useful to determine the best path forward.

The annoying situation here is that the code only ever got accepted because the compiler was extended without RFCs or design/planning... and now we realize that the design that makes most sense accepts less code than what the compiler accepted through these accidents.

Maybe we should wait until inline const expressions support surrounding generics (@oli-obk any timeline for that?).

> So is there an easy way to get the nightly compiler with these changes on x86-64 linux to try making those changes in abi_stable?

Depends on your notion of "easy". ;)
With https://github.com/kennytm/rustup-toolchain-install-master installed, you can do `rustup-toolchain-install-master 81275a1445c7502a974118e0c69b795ab60be991`.

**Comment by oli-obk** [comment]
> Maybe we should wait until inline const expressions support surrounding generics (@oli-obk any timeline for that?).

We can probably start experimenting with that quickly, as making them support surrounding generics is unproblematic when compared to allowing surrounding generics in array lengths.

**Comment by rodrimati1992** [comment]
I've downloaded the rustc from https://github.com/rust-lang/rust/pull/80243.

Starting with the changes to `abi_stable` that are entirely internal, looks like making `abi_stable` compile with these changes to promotion would take multiple days, since I pervasively use promotion of the parameters/return values of `const fns` in macros, in ways that would require not only rewriting the macros but also every use of those macros.

It also would require making a breaking change to an attribute (requiring the user to pass the type of the expression passed to the  [`#[sabi(extra_checks = "<expression>")]`](https://docs.rs/abi_stable/0.9.1/abi_stable/docs/stable_abi_derive/index.html#sabiextra_checkssome_constant_expression) helper attribute).

**Comment by rodrimati1992** [comment]
This comment is about the changes made to make abi_stable compile with  [the pull request that disables promotion of all const fn calls](https://github.com/rust-lang/rust/pull/80243)

So I worked on getting most of the code working between sunday and monday (in the [fix_fn_promotion branch](https://github.com/rodrimati1992/abi_stable_crates/tree/fix_fn_promotion)), it wasn't too difficult, but it took a while.


# Changes to make code work

The relevant changes I needed to make code work were:

- Moving borrows that used to be promoted into (associated) `const`s.

- Move the elements passed to the  [`rslice`] macro into a separate `&[_]` `const` item, and replace the [`rslice`] macro invocation with [`RSlice::from_slice`] called with the slice constant.


# Broken API

The API that were broken (described in sections below) can mostly be fixed with inline consts (`const {}`) that can refer to generic  parameters in scope, used internally by the macros.

While this is not something that will affect me, using inline consts for promotion inside the [`tag`] and [`rslice`] macros would be breaking for anyone who accesses local variables in the macro invocations.

### `tag` macro

The [`tag`] macro allows constructing json-like constants of the `Tag` type.

The pull request breaks the [`tag`] macro, by causing the input elements to dangle if any element expression contains a function call or a nested invocation of the [`tag`] macro. 

Before the pull request, the tag macro worked in a tail expression to initialize a `const` or `static`.

These are documentation examples that break in the [tagging] module:
- [Declaring each variant](https://docs.rs/abi_stable/0.9.1/abi_stable/type_layout/tagging/index.html#declaring-each-variant)
- [Creating a complex data structure](https://docs.rs/abi_stable/0.9.1/abi_stable/type_layout/tagging/index.html#creating-a-complex-data-structure)

I intend Tag to also be usable with the [`StableAbi` derive] through the [`#[sabi(tag = "....")]` helper attribute](https://docs.rs/abi_stable/0.9.1/abi_stable/docs/stable_abi_derive/index.html#sabitag-some_expr-), where it has to be able to refer to associated constants and (now) const parameters.


### `rslice` macro

The pull request breaks the [`rslice`] macro, 
by causing the input elements to dangle if any element expression contains a function call.

Before the pull request, the rslice macro worked in a tail expression to initialize a `const` or `static`.

Code that uses [`rslice`] like this normally works, and doesn't with hte pull request:
```rust
use abi_stable::{
    std_types::RSlice,
    rslice,
};

const fn foo() -> u8 {
    100
}

const _: RSlice<'_, u8> = rslice![0, 1, foo()];
```
I had to refactor all code like that into:
```rust
const S: &[u8] = &[0, 1, foo()];
const _: RSlice<'_, u8> = RSlice::from_slice(S);
```
effectively making the macro ususable for non-literal elements.


[`rslice`]: https://docs.rs/abi_stable/0.9.1/abi_stable/macro.rslice.html
[`tag`]: https://docs.rs/abi_stable/0.9.1/abi_stable/macro.tag.html
[tagging]: https://docs.rs/abi_stable/0.9.1/abi_stable/type_layout/tagging/index.html
[`StableAbi` derive]: https://docs.rs/abi_stable/0.9.1/abi_stable/derive.StableAbi.html
[`RSlice::from_slice`]: https://docs.rs/abi_stable/0.9.1/abi_stable/std_types/struct.RSlice.html#method.from_slice

**Comment by oli-obk** [comment]
Thank you for this in depth analysis! I'm not sure what our next steps are, we are just figuring out things on our end. It is still possible we'll never end up changing the rules

**Comment by RalfJung** [comment]
Thank you indeed. :-)

> While this is not something that will affect me, using inline consts for promotion inside the tag and rslice macros would be breaking for anyone who accesses local variables in the macro invocations.

Could you explain this in some more detail? Promotion will also not apply if local variables are used in the expression, so I don't think using `const {}` instead can possibly break anything.

**Comment by rodrimati1992** [comment]
@RalfJung 
This is currently valid:
```rust
const fn compute_u8() -> u8 {
    10
}

const FOO: &[u8] = {
    let number = compute_u8();
    
    &[number, number]
};
```
If `rslice` macro changed to use `const { &[$($elem),*] }` to promote the array, this:
```rust
use abi_stable::std_types::RSlice;
use abi_stable::rslice;

const fn compute_u8() -> u8 {
    10
}

const FOO: RSlice<'_, u8> = {
    let number = compute_u8();
    
    rslice![number, number]
};
```
would not compile, because AFAIK inline consts cannot refer to local variables.

**Comment by RalfJung** [comment]
That is true. However, the tail expression rule isn't changing, so you shouldn't have to use `const {...}` here even with my PR.

**Comment by RalfJung** [comment]
Or, to use terminology from [the promotion documentation](https://github.com/rust-lang/const-eval/blob/master/promotion.md) -- your `FOO` (in the first example) does not use promotion. It uses the "enclosing scope rule", and no amount of changes to promotion will make it stop working.
